Standardize string formatting to use f-strings

tnwei opened the pull request:
As per #270 , submitting this pull request to revamp string formatting in `pdoc` to use f-strings. Also bumped min requirements of Python 3.5+ to Python 3.6+ in `setup.py`.

Following `CONTRIBUTING.md`:
+ All tests passed from `python setup.py test`
+ Ran `flake8 --max-line-length 100`, no new warnings. At time of writing, existing warning is:
```
 ./pdoc/test/example_pkg/_resolve_typing_forwardrefs/postponed.py:1:1: F407 future feature annotations is not defined
```

kernc commented on pdoc/__init__.py:
```diff
@@ -776,7 +776,7 @@ def _link_inheritance(self):
                 continue
             if not isinstance(docstring, str):
                 raise ValueError('__pdoc__ dict values must be strings;'
-                                 '__pdoc__[{!r}] is of type {}'.format(name, type(docstring)))
+                                 f'__pdoc__[{name!r}] is of type {type(docstring)}')
```
Missing space at the end of first line.

tnwei replied:
Can you please elaborate where is the missing space? I've checked that the before/after edit prints the same line

kernc replied:
> ```py
> '__pdoc__ dict values must be strings; '
> ```

Was missing before already. :sweat_smile:

tnwei replied:
Ah, got it.

kernc commented on pdoc/cli.py:
```diff
@@ -236,8 +236,8 @@ def do_GET(self):
-                out = "Error importing module <code>{}</code>:\n\n<pre>{}</pre>".format(
-                    self.import_path_from_req_url, escape(traceback.format_exc()))
+                out = f"Error importing module <code>{self.import_path_from_req_url}</code>:"
+                f"\n\n<pre>{escape(traceback.format_exc())}</pre>"
```
Indent, wrap in parentheses.

tnwei replied:
Something like this looks better?
``` python
out = (
    "Error importing module "
    f"<code>{self.import_path_from_req_url}</code>:\n\n"
    f"<pre>{escape(traceback.format_exc())}</pre>"
)
```

kernc commented on pdoc/__init__.py:
```diff
@@ -140,9 +140,10 @@ def _render_template(template_name, **kwargs):
-                ', '.join(path.join(p, template_name.lstrip("/"))
-                          for p in tpl_lookup.directories)))
+            "No template found at any of: "
+            f"{', '.join(path.join(p, template_name.lstrip('/')) for p in tpl_lookup.directories)}"
```
Maybe just `"one" + "other"` or a new `paths` variable?

tnwei replied:
Yeah that looks better, how about the following?
```python
paths = (path.join(p, template_name.lstrip('/')) for p in tpl_lookup.directories)
raise OSError(f"No template found at any of: {', '.join(paths)}")

kernc reviewed (commented):
Note, _pdoc/html_helpers.py_ and _pdoc/templates/{html,pdf}.mako_ contain some further `str.format()` / `str.__mod__` calls.

tnwei commented:
> Note, pdoc/html_helpers.py and pdoc/templates/{html,pdf}.mako contain some further str.format() / str.__mod__ calls.

You are right, will be checking on the missed files

tnwei commented:
Re-checked all documents to pick up on missed `__mod__` and `.format()` calls:
+ [X] `pdoc/__init__.py`
+ [X] pdoc/cli.py
+ [X] `pdoc/test/__init__.py`
+ [ ] pdoc/html_helpers.py (line 176)
+ [X] pdoc/templates/html.mako
+ [ ] pdoc/templates/pdf.mako (line 18)

tnwei commented:
Two questions:

#### pdoc/html_helpers.py (line 176)
`str.format` takes output from a `map` nested in list comprehension as below:
``` python
return '\n\n'.join('`{}`\n:   {}'.format(*map(str.strip, line.split(':', 1)))
                       for line in filter(None, spec_with_desc.split('\n')))
```

Reformatting it would give
``` python
return '\n\n'.join(f'`{line.split(":", 1)[0].strip()}`\n:   '
            f'{line.split(":", 1)[1].strip()}' for line in filter(None, spec_with_desc.split('\n')))
```
but it looks rather inelegant. Are we good with this?

#### pdoc/templates/pdf.mako (line 18)

``` python
text = re.sub(r'^(?P<indent>( *))!!! \w+ \"([^\"]*)\"(.*(?:\n(?P=indent) +.*)*)',
                      lambda m: '{}**{}:** {}'.format(m.group(2), m.group(3),
                                                      re.sub('\n {,4}', '\n', m.group(4))),
                      text, flags=re.MULTILINE)
```

One of the inputs to `str.format` above include evaluating regex substitution where newline chars are involved. f-strings don't allow backslash in the curly braces, and assigning it to a variable will need rewriting the lambda function that is passed to `re.sub()`, like so:
``` python
def paragraph_fmt(m):
    sub = re.sub('\n {,4}', '\n', m.group(4))
    return f'{m.group(2)}**{m.group(3)}:** {sub}'
text = re.sub(r'^(?P<indent>( *))!!! \w+ \"([^\"]*)\"(.*(?:\n(?P=indent) +.*)*)',
                     paragraph_fmt, text, flags=re.MULTILINE)
```

Would love to have your comments @kernc

kernc commented:
Snippets like these show `str.format/%` still has some value. And how awful regexes are.

I don't mind new variables in both cases. In the first case maybe:
```py
spec, desc = map(...)
return f"..."
```
When in doubt, opt for clarity.

tnwei commented:
> Snippets like these show str.format/% still has some value. And how awful regexes are.

Thanks for the thoughts, at some point I was contemplating if it's better to leave them as-is actually. 

Anywho, I've finished updating each file in their own commit, whole PR is ready for review.

kernc commented on pdoc/html_helpers.py:
```diff
@@ -277,13 +280,13 @@ def _admonition(match, module=None, limit_types=None):
-                indent, text.translate(str.maketrans({'\n': ' ',
-                                                      '[': '\\[',
-                                                      ']': '\\]'})).strip(), value)
+            strfmt = text.translate(str.maketrans({
+                '\n': ' ',
+                '[': '\\[',
+                ']': '\\]'})).strip()
+            return f'{indent}![{strfmt}]({value})\n'
```
```suggestion
            alt_text = text.translate(str.maketrans({
                '\n': ' ',
                '[': '\\[',
                ']': '\\]'})).strip()
            return f'{indent}![{alt_text}]({value})\n'
```

kernc commented:
Many thanks @tnwei! Hope you got your t(r)ee. :smile:
